hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/IncrementalRelation.scala

```diff
@@ -90,24 +78,24 @@ class IncrementalRelation(val sqlContext: SQLContext,
     throw new HoodieException("Incremental queries are not supported when meta fields are disabled")
   }
 
+  private val queryContext: IncrementalQueryAnalyzer.QueryContext =
+    IncrementalQueryAnalyzer.builder()
+      .metaClient(metaClient)
+      .startTime(optParams(DataSourceReadOptions.BEGIN_INSTANTTIME.key))
+      .endTime(optParams.getOrElse(DataSourceReadOptions.END_INSTANTTIME.key, null))
+      .rangeType(InstantRange.RangeType.OPEN_CLOSED)
```

Comment:
Why the start instant is hard code as ignored, for continuous streaming read, it may make sense but may not for batch incremental source.

Reply:
In Spark the `hoodie.datasource.read.begin.instanttime` is exclusive, being treated as the last checkpoint just like `HoodieStreamer`'s last checkpoint of the source.

Reply:
I've changed `IncrSourceHelper#getIncrementalQueryAnalyzer` to use different `RangeType` based on whether there is last checkpoint or not, so `CLOSED_CLOSED` is used for the first batch, while `OPEN_CLOSED` is used for subsequent batches.

Reply:
Got it, why we need a separate query context for `IncrementalRelation` and `MergeOnReadIncrementalRelation` with hard code range type as `OPEN_CLOSED`, should we make the range type configurable for these two?

Reply:
I've fixed that to be `CLOSED_CLOSED` in my latter git commits.  The Hudi incremental source or stream source first determines the range of instants though the incremental query analyzer with either `OPEN_CLOSED` range for resuming from the last checkpoint, or `CLOSED_CLOSED` range for the first batch without checkpoint.  Then the begin and end completion is passed down to the relation with both as inclusive.

Reply:
This is fixed now by dynamically setting the range type for the incremental query.